README.md
# ZipUtilities
[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/NSProgrammer/ZipUtilities/master/LICENSE.md) [![CocoaPods compatible](https://img.shields.io/cocoapods/v/ZipUtilities.svg?style=flat)](https://CocoaPods.org) [![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)

![Targets](https://img.shields.io/badge/Targets-OSX.Framework_iOS.Framework_iOS.lib-lightgrey.svg)

[![Build Status](https://travis-ci.org/NSProgrammer/ZipUtilities.svg?branch=master)](https://travis-ci.org/NSProgrammer/ZipUtilities)

## Introduction

*ZipUtilities*, prefixed with `NOZ` for _Nolan O'Brien ZipUtilities_, is a library of zipping and unzipping utilities for iOS and Mac OS X.

## Background

The need can occasionally arise where easy compressing and decompressing of data is desired from a simple API.  There are many zipping/unzipping utilities out there but all of them can be found wanting in some way or another.

- too low level
- too complex
- poor code quality
- old coding practices/style (basically really old)
- lack a service oriented approach (request, operation and response pattern)

The goal is to provide an easy to use modern interface for archiving and unarchiving zip files. As a particular focus, providing a service oriented approach can provide powerful support for NSOperation composition.

## Install

The _ZipUtilities_ Xcode project has targets to build iOS and OSX dynamic frameworks. You can build these and add them to your project manually, or add a subproject in Xcode.

Alternatively you may use one of the following dependency managers:

#### CocoaPods

Add _ZipUtilities_ to your `Podfile`

```ruby
pod 'ZipUtilities', '~> 1.13.0'
```

#### Carthage

Add _ZipUtilities_ to your `Cartfile`

```ruby
github "NSProgrammer/ZipUtilities"
```

## Documentation

You can either build the documentation with appledoc locally or you can visit the documentation online at http://cocoadocs.org/docsets/ZipUtilities

## Overview

### Service Oriented Interfaces (NSOperations)

The primary value of _ZipUtilities_ is that it provides an easy to use interface for archiving data or files into a single zip archive and unarchiving a zip archive to the contained files.  The primary approach for _ZipUtilities_ is to provide a service oriented pattern for compressing and decompressing.

**`NOZCompress.h`**

`NOZCompress.h` contains the service oriented interfaces related to compressing into a zip archive.

- `NOZCompressRequest` is the object that encapsulates the _what_ and _how_ for the compression operation to act upon
- `NOZCompressOperation` is the `NSOperation` subclass object that performs the compression. By being an `NSOperation`, consumers can take advantage of cancelling, prioritization and dependencies.  Progress is also provided with the operation and can be observed via _KVO_ on the `progress` property or via the delegate callback.
- `NOZCompressDelegate` is the delegate for the `NOZCompressOperation`. It provides callbacks for progress and completion.
- `NOZCompressResult` is the object that encapsulates the result of a compress operation. It holds whether or not the operation succeed, the error if it didn't succeed, the path to the created zip archive and other informative metrics like duration and compression ratio.

*Example:*

```obj-c
- (NSOperation *)startCompression
{
	NOZCompressRequest *request = [[NOZCompressRequest alloc] initWithDestinationPath:self.zipFilePath];
    [request addEntriesInDirectory:self.sourceDirectoryPath 
                       filterBlock:^BOOL(NSString *filePath) {
        return [filePath.lastPathComponent hasPrefix:@"."];
    }
        compressionSelectionBlock:NULL];
    [request addDataEntry:self.data name:@"Aesop.txt"];
    for (id<NOZZippableEntry> entry in self.additionalEntries) {
        [request addEntry:entry];
    }

    NOZCompressionOperation *op = [[NOZCompressOperation alloc] initWithRequest:request delegate:self];
    [self.operationQueue addOperation:op];

    // return operation so that a handle can be maintained and cancelled if necessary
    return op;
}

- (void)compressOperation:(NOZCompressOperation *)op didCompleteWithResult:(NOZCompressResult *)result
{
    dispatch_async(dispatch_get_main_queue(), ^{
	    self.completionBlock(result.didSuccess, result.operationError);
    });
}

- (void)compressOperation:(NOZCompressOperation *)op didUpdateProgress:(float)progress
{
	dispatch_async(dispatch_get_main_queue(), ^{
	    self.progressBlock(progress);
	});
}
```

```swift
func startCompression() -> NSOperation
{
    let request = NOZCompressRequest.init(destinationPath: self.zipFilePath)
    request.addEntriesInDirectory(self.sourceDirectoryPath, filterBlock: { (filePath: String) -> Bool in
        return ((filePath as NSString).lastPathComponent as NSString).hasPrefix(".")
    }, compressionSelectionBlock: nil)
    request.addDataEntry(self.data name:"Aesop.txt")
    for entry in self.additionalEntries {
        request.addEntry(entry)
    }

    let operation = NOZCompressOperation.init(request: request, delegate: self)
    zipQueue?.addOperation(operation)
    
    // return operation so that a handle can be maintained and cancelled if necessary
    return operation
}

func compressOperation(op: NOZCompressOperation, didCompleteWithResult result: NOZCompressResult)
{
    dispatch_async(dispatch_get_main_queue(), {
        self.completionBlock(result.didSuccess, result.operationError);
    })
}

func compressOperation(op: NOZCompressOperation, didUpdateProgress progress: Float)
{
    dispatch_async(dispatch_get_main_queue(), {
        self.progressBlock(progress);
    })
}
```

**`NOZDecompress.h`**

`NOZDecompress.h` contains the service oriented interfaces related to decompressing from a zip archive.

- `NOZDecompressRequest` is the object that encapsulates the _what_ and _how_ for the decompression operation to act upon
- `NOZDecompressOperation` is the `NSOperation` subclass object that performs the compression. By being an `NSOperation`, consumers can take advantage of cancelling, prioritization and dependencies.  Progress is also provided with the operation and can be observed via _KVO_ on the `progress` property or via the delegate callback.
- `NOZDecompressDelegate` is the delegate for the `NOZDecompressOperation`. It provides callbacks for progress, overwriting output files and completion.
- `NOZDecompressResult` is the object that encapsulates the result of a compress operation. It holds whether or not the operation succeed, the error if it didn't succeed, the paths to the output unarchived files and other informative metrics like duration and compression ratio.

*Example:*

```obj-c
- (NSOperation *)startDecompression
{
    NOZDecompressRequest *request = [[NOZDecompressRequest alloc] initWithSourceFilePath:self.zipFilePath];

    NOZDecompressOperation *op = [[NOZDecompressOperation alloc] initWithRequest:request delegate:self];
    [self.operationQueue addOperation:op];

    // return operation so that a handle can be maintained and cancelled if necessary
    return op;
}

- (void)decompressOperation:(NOZDecompressOperation *)op didCompleteWithResult:(NOZDecompressResult *)result
{
    dispatch_async(dispatch_get_main_queue(), ^{
	    self.completionBlock(result.didSuccess, result.destinationFiles, result.operationError);
    });
}

- (void)decompressOperation:(NOZDecompressOperation *)op didUpdateProgress:(float)progress
{
	dispatch_async(dispatch_get_main_queue(), ^{
	    self.progressBlock(progress);
	});
}
```

```swift
func startDecompression() -> NSOperation
{
    let request = NOZDecompressRequest.init(sourceFilePath: self.zipFilePath)
    let operation = NOZDecompressOperation.init(request: request, delegate: self)
    zipQueue?.addOperation(operation)
    return operation
}

func decompressOperation(op: NOZDecompressOperation, didCompleteWithResult result: NOZDecompressResult)
{
    dispatch_async(dispatch_get_main_queue(), {
        self.completionBlock(result.didSuccess, result.destinationFiles, result.operationError);
    })
}

func decompressOperation(op: NOZDecompressOperation, didUpdateProgress progress: Float)
{
    dispatch_async(dispatch_get_main_queue(), {
        self.progressBlock(progress);
    })
}
```

### Manual Zipping and Unzipping

Additional, the underlying objects for zipping and unzipping are exposed for direct use if NSOperation support is not needed.

**`NOZZipper.h`**

`NOZZipper` is an object that encapsulates the work for zipping sources (NSData, streams and/or
files) into an on disk zip archive file.

*Example:*

```obj-c
- (BOOL)zipThingsUpAndReturnError:(out NSError **)error
{
    NOZZipper *zipper = [[NOZZipper alloc] initWithZipFile:pathToCreateZipFile];
    if (![zipper openWithMode:NOZZipperModeCreate error:error]) {
        return NO;
    }

    __block int64_t totalBytesCompressed = 0;

    NOZFileZipEntry *textFileZipEntry = [[NOZFileZipEntry alloc] initWithFilePath:textFilePath];
    textFileZipEntry.comment = @"This is a heavily compressed text file.";
    textFileZipEntry.compressionLevel = NOZCompressionLevelMax;

    NSData *jpegData = UIImageJPEGRepresentation(someImage, 0.8f);
    NOZDataZipEntry *jpegEntry = [[NOZDataZipEntry alloc] initWithData:jpegData name:@"image.jpg"];
    jpegEntry.comment = @"This is a JPEG so it doesn't need more compression.";
    jpegEntry.compressionMode = NOZCompressionModeNone;

    if (![zipper addEntry:textFileZipEntry
            progressBlock:^(int64_t totalBytes, int64_t bytesComplete, int64_t bytesCompletedThisPass, BOOL *abort) {
            totalBytesCompressed = bytesCompletedThisPass;
          }
                    error:error]) {
        return NO;
    }

    if (![zipper addEntry:jpegEntry
            progressBlock:^(int64_t totalBytes, int64_t bytesComplete, int64_t bytesCompletedThisPass, BOOL *abort) {
            totalBytesCompressed = bytesCompletedThisPass;
         }
                    error:error]) {
        return NO;
    }

    zipper.globalComment = @"This is a global comment for the entire archive.";
    if (![zipper closeAndReturnError:error]) {
        return NO;
    }

    int64_t archiveSize = (int64_t)[[[NSFileManager defaultFileManager] attributesOfItemAtPath:zipper.zipFilePath] fileSize];
    NSLog(@"Compressed to %@ with compression ratio of %.4f:1", zipper.zipFilePath, (double)totalBytesCompressed / (double)archiveSize);
    return YES;
}

```

**`NOZUnzipper.h`**

`NOZUnzipper` is an object that encapsulates the work for unzipping from a zip archive file on disk
into destinations (NSData, streams and/or files).

*Example:*

```obj-c
- (BOOL)unzipThingsAndReturnError:(out NSError **)error
{
    NSAssert(![NSThread isMainThread]); // do this work on a background thread

    NOZUnzipper *unzipper = [[NOZUnzipper alloc] initWithZipFile:zipFilePath];
    if (![unzipper openAndReturnError:error]) {
        return NO;
    }

    if (nil == [unzipper readCentralDirectoryAndReturnError:error]) {
        return NO;
    }

    __block NSError *enumError = nil;
    [unzipper enumerateManifestEntriesUsingBlock:^(NOZCentralDirectoryRecord * record, NSUInteger index, BOOL * stop) {
        NSString *extension = record.name.pathExtension;
        if ([extension isEqualToString:@"jpg"]) {
            *stop = ![self readImageFromUnzipper:unzipper withRecord:record error:&enumError];
        } else if ([extension isEqualToString:@"json"]) {
            *stop = ![self readJSONFromUnzipper:unzipper withRecord:record error:&enumError];
        } else {
            *stop = ![self extractFileFromUnzipper:unzipper withRecord:record error:&enumError];
        }
    }];

    if (enumError) {
        *error = enumError;
        return NO;
    }

    if (![unzipper closeAndReturnError:error]) {
        return NO;
    }

    return YES;
}

- (BOOL)readImageFromUnzipper:(NOZUnzipper *)unzipper withRecord:(NOZCentralDirectoryRecord *)record error:(out NSError **)error
{
    CGImageSourceRef imageSource = CGImageSourceCreateIncremental(NULL);
    custom_defer(^{ // This is Obj-C equivalent to 'defer' in Swift.  See http://www.openradar.me/21684961 for more info.
        if (imageSource) {
            CFRelease(imageSource);
        }
    });

    NSMutableData *imageData = [NSMutableData dataWithCapacity:record.uncompressedSize];
    if (![unzipper enumerateByteRangesOfRecord:record
                                 progressBlock:NULL
                                    usingBlock:^(const void * bytes, NSRange byteRange, BOOL * stop) {
                                        [imageData appendBytes:bytes length:byteRange.length];
                                        CGImageSourceUpdate(imageSource, imageData, NO);
                                    }
                                         error:error]) {
        return NO;
    }

    CGImageSourceUpdate(imageSource, (__bridge CFDataRef)imageData, YES);
    CGImageRef imageRef = CGImageSourceCreateImageAtIndex(imageSource, 0, NULL);
    if (!imageRef) {
        *error = ... some error ...;
        return NO;
    }

    custom_defer(^{
        CFRelease(imageRef);
    });

    UIImage *image = [UIImage imageWithCGImage:imageRef];

CHANGELOG.md
# ZipUtilities Change Log

## History

### 1.13.0 (June 18th, 2021) - Nolan O'Brien
- Update ZStandard extended support to v1.5.0
- Update constants of spec based compression methods (including addition of ZStandard as #93)

### 1.12.0 (June 18th, 2021) - Nolan O'Brien
- Fix benign decoder bug where the incorrect compression level could be inferred
- Fix significant bug in string encoding support
  - Need to handle both UTF8 encoded filenames/comments AND DOS Latin US (Code Page 437) encoded (Appendix D of ZIP Specification)
  - ***Note:** Any zip archives created with ZipUtilities **v1.11.3 or older** that have non-ascii characters for filenames/comments will have mangled filenames/comments as they will decode using DOS Latin US character set instead of UTF8.*

### 1.11.3 (July 30, 2020) - Nolan O'Brien
- Update Brotli (v1.0.7) and ZStandard (v1.4.5)
- Other minor fixes

### 1.11.2 (September 4, 2018) - Nolan O'Brien
- Fix vulnerability when zip archive provides malicious output file name

### 1.11.1 (August 14, 2018) - Nolan O'Brien

- Update Brotli (v1.0.5) and ZStandard (v1.3.5)
- Bug fixes for CLI 

### 1.11.0 (April 24, 2018) - Nolan O'Brien

- Update Brotli (v1.0.4) and ZStandard (v1.3.4)
  - eliminates shared custom dictionary support from brotli

### 1.10.1 (April 23, 2018) - Nolan O'Brien

- Add _Zip Mode_ support to `noz`
- Fix `noz` bugs

### 1.10.0 (June 3, 2017) - Nolan O'Brien

- add convenience file-to-file compression/decompression functions in `NOZUtils.h`
- add CLI for _ZipUtilities_ (called `noz`)
    - NOTE: Zip Mode is not yet implemented
- fix `NOZUnzipper` file size measurement bug

### 1.9.3 (Feb 20, 2017) - Nolan O'Brien

- add checksum error code and record validation method to unzipper

### 1.9.2 (Feb 19, 2017) - Nolan O'Brien

- updated zstd to v1.1.3

### 1.9.1 (Jan 31, 2017) - Nolan O'Brien

- Miscellaneous minor fixes

### 1.9.0 (Nov 22, 2016) - Nolan O'Brien

- Add Brotli support as an extra encoder/decoder
- Refactor compression level
  - decouples from ZLib levels (1-9) and moves to a float range (from 0.0 to 1.0)
  - can map to any encoder with multiple levels now
- Fix memory leak in Apple extra encoders/decoders
- Bump up buffer sizes from 4KB to 16KB
- Update unit tests
- Update test app
- rearrange some code paths for better 3rd party codec support
- improve memory management to reduce memory footprint when multiple encodes/decodes happen

### 1.8.1 (Nov 18, 2016) - Nolan O'Brien

- Fix up project

### 1.8.0  (Nov 17, 2016) - Nolan O'Brien

- Add ZStandard support as an extra encoder/decoder
- Clean up some files
- Fix an edge case in NSData+NOZAdditions category
- Clean up schemes
- Add codec comparison unit test
- Add ZipUtilitiesApp iOS app for testing codec perf on device

### 1.7.2  (Jan 11, 2016) - Nolan O'Brien

- Provide options when saving a record so that we can support writing an entry to disk without the interim path

### 1.7.1  (Dec 31, 2015) - Nolan O'Brien

- Add ability to filter files being added to an NOZCompressRequest when adding files via containing directory path

### 1.7.0  (Dec 15, 2015) - Nolan O'Brien

- Fix NOZErrorCodes (paging was off)
- Fix Decoding large files with Deflate
- Added unit tests

### 1.6.6  (Oct 6, 2015) - Nolan O'Brien

- Fix bug in Unzipper when "overwrite" is NO
- Add Swift unit tests
- Add Swift example code for operation based compression/decompression
- Fix compression ratio of NOZDecompressResult

### 1.6.5  (Oct 4, 2015) - Nolan O'Brien

- Fix bug in Apple Encoder/Decoder
- Add convenience NSInputStream for compressed streams (could be optimized further)

### 1.6.0  (Sep 26, 2015) - Nolan O'Brien

- Rename NOZCompressionEncoder/Decoder to NOZEncoder/Decoder
- Add category to NSData for easy compression/decompression of data

### 1.5.1  (Sep 13, 2015) - Nolan O'Brien

- Update docs
- Minor cleanup

### 1.5.0  (Sep 11, 2015) - Nolan O'Brien

- Simplify compression encoders/decoders to only return a BOOL and not an NSError
- Convert project to Xcode 7
- Use container generics throughout
- Add LZMA encoder/decoder to "extras"
  - Includes unit tests
  - iOS 9+ and OS X 10.11+ only
  - Not included in libs/frameworks, you can include these files to add LZMA support though

### 1.4.1  (Sep 6, 2015) - Nolan O'Brien

- Fix race condition with cancelling
- Change runStep: method to runStep:error: for better consistency and Swift compatibility

### 1.4.0  (Sep 5, 2015) - Ashton Williams

- Mac OS X target and support
- iOS dynamic framework target

### 1.3.2  (Aug 23, 2015) - Nolan O'Brien

- Optimize zipping to be done in a single write pass
  - Can be disabled with NOZ_SINGLE_PASS_ZIP being set to 0

### 1.3.1  (Aug 23, 2015) - Nolan O'Brien

- Finish modularization of compression with decoders being implemented

### 1.3.0  (Aug 22, 2015) - Nolan O'Brien

- Modularize compression encoding and decoding with protocols in `NOZCompression.h`
- Add _Deflate_ encoder/decoder
- Add _Raw_ encoder/decoder (for no compression support)
- TODO: currently, only encoders are supported and implemented. Still need to implement and use decoders.

### 1.2.0  (Aug 21, 2015) - Nolan O'Brien

- Implement NOZUnzipper
- Use NOZUnzipper for decompression
- Remove minizip dependency

### 1.1.0  (Aug 15, 2015) - Nolan O'Brien

- Introduce NOZZipper for compression
- Remove minizip dependency for compression
- TODO: implement NOZUnzipper and remove minizip dependency completely
- NOTE: loses support for password encryption and ZIP64 support

### 1.0.1b (Aug 11, 2015) - Nolan O'Brien

- Straighten out some minizip code

### 1.0.1 (Aug 10, 2015) - Nolan O'Brien

- Added comments/documentation throughout headers
- Added more places to detect cancellation to NOZDecompressOperation

### 1.0.0 (Aug 9, 2015) - Nolan O'Brien

- Refactored code
- fixed compilation and static analysis warnings in minizip 1.1
- pushed to github ( https://github.com/NSProgrammer/ZipUtilities )

### 0.9.0 (Aug 4, 2015) - Nolan O'Brien


noz/NOZCLI.m
//
//  NOZCLI.m
//  ZipUtilities
//
//  Created by Nolan O'Brien on 5/7/17.
//  Copyright © 2017 NSProgrammer. All rights reserved.
//

#import "NOZCLI.h"

#import "NOZCLICompressMode.h"
#import "NOZCLIDecompressMode.h"
#import "NOZCLIDumpMode.h"
#import "NOZCLIMethodMode.h"
#import "NOZCLIUnzipMode.h"
#import "NOZCLIZipMode.h"

#import "NOZXAppleCompressionCoder.h"
#import "NOZXBrotliCompressionCoder.h"
#import "NOZXZStandardCompressionCoder.h"

#define kMethodBrotli       (101)

@implementation MethodInfo

+ (instancetype)methodInfoWithName:(NSString *)name
                            method:(NOZCompressionMethod)method
                            levels:(NSUInteger)levels
                      defaultLevel:(NSUInteger)defaultLevel
                           encoder:(id<NOZEncoder>)encoder
                           decoder:(id<NOZDecoder>)decoder
                         isDefault:(BOOL)defaultEncoder
{
    return [[self alloc] initWithName:name
                               method:method
                               levels:levels
                         defaultLevel:defaultLevel
                              encoder:encoder
                              decoder:decoder
                            isDefault:defaultEncoder];
}

- (instancetype)initWithName:(NSString *)name
                      method:(NOZCompressionMethod)method
                      levels:(NSUInteger)levels
                defaultLevel:(NSUInteger)defaultLevel
                     encoder:(id<NOZEncoder>)encoder
                     decoder:(id<NOZDecoder>)decoder
                   isDefault:(BOOL)defaultEncoder
{
    if (self = [super init]) {
        _name = [name copy];
        _method = method;
        _levels = levels;
        _defaultLevel = defaultLevel;
        _defaultCodec = defaultEncoder;
        _encoder = encoder;
        _decoder = decoder;
    }
    return self;
}

- (NSString *)description
{
    NSString *codecInfo = nil;
    if (_method == NOZCompressionMethodNone) {
        codecInfo = @"Default";
    } else if (!_encoder || !_decoder) {
        codecInfo = @"Unsupported";
    } else if (_defaultCodec) {
        codecInfo = @"Default";
    } else {
        codecInfo = @"Extended";
    }

    return [NSString stringWithFormat:@"<%@: %p, name=\"%@\", methodEnum=%u, levels=%tu, codec=%@>", NSStringFromClass([self class]), self, _name, (unsigned int)_method, _levels, codecInfo];
}

@end

#define METHOD(theName, theMethod, theLevels, theDefaultLevel, theEncoder, theDecoder, theIsDefault) \
[MethodInfo methodInfoWithName:(theName) method:(theMethod) levels:(theLevels) defaultLevel:(theDefaultLevel) encoder:(theEncoder) decoder:(theDecoder) isDefault:(theIsDefault)]

#define ADD_METHOD(theArray, theName, theMethod, theIsDefault) \
do { \
    NOZCompressionMethod the_method = (theMethod); \
    id<NOZEncoder> the_encoder = [[NOZCompressionLibrary sharedInstance] encoderForMethod:the_method]; \
    id<NOZDecoder> the_decoder = [[NOZCompressionLibrary sharedInstance] decoderForMethod:the_method]; \
    if (!the_encoder || !the_decoder) { \
        the_encoder = nil; \
        the_decoder = nil; \
    } \
    NSUInteger the_levels = [the_encoder respondsToSelector:@selector(numberOfCompressionLevels)] ? [the_encoder numberOfCompressionLevels] : 0; \
    NSUInteger the_defaultLevel = [the_encoder respondsToSelector:@selector(defaultCompressionLevel)] ? [the_encoder defaultCompressionLevel] : 0; \
    [(theArray) addObject:METHOD((theName), the_method, the_levels, the_defaultLevel, the_encoder, the_decoder, (theIsDefault))]; \
} while (0)

#define UNS_ADD_METHOD(theArray, theName, theMethod) \
do { \
    NOZCompressionMethod the_method = (theMethod); \
    id<NOZEncoder> the_encoder = [[NOZCompressionLibrary sharedInstance] encoderForMethod:the_method]; \
    id<NOZDecoder> the_decoder = [[NOZCompressionLibrary sharedInstance] decoderForMethod:the_method]; \
    if (the_encoder && the_decoder) { \
        break; \
    } \
    [(theArray) addObject:METHOD((theName), the_method, 0, 0, nil, nil, NO)]; \
} while (0)

int NOZCLI_main(NSString *exe,
                NSString *exeDir,
                NSString *currentDir,
                NSArray<NSString *> *args)
{
    if (args.count == 0) {
        return -1;
    }

    NOZCLI_registerExtraEncoders();

    NSString *modeFlag = args[0];
    args = [args subarrayWithRange:NSMakeRange(1, args.count - 1)];

    NSArray<Class> *modes = NOZCLI_allModes();

    for (Class mode in modes) {
        if ([modeFlag isEqualToString:[mode modeFlag]]) {
            id<NOZCLIModeInfoProtocol> info = [mode infoFromArgs:args environmentPath:currentDir];
            return [mode run:info];
        }
    }

    return -1;
}

void NOZCLI_printUsage(NSString *exe, NSString *modeFlag)
{
    if (exe) {
        exe = @"noz";
    }

    printf("%s is the ZipUtilities CLI\n\n", exe.UTF8String);

    NSArray<Class> *modes = NOZCLI_allModes();

    if (modeFlag) {
        // if a mode was selected, just print how to use that mode
        for (Class mode in modes) {
            if ([[mode modeFlag] isEqualToString:modeFlag]) {
                modes = @[mode];
                break;
            }
        }
    }

    for (Class mode in modes) {
        NSString *modeDescription = [NSString stringWithFormat:@"%@ mode:\n\t%@ %@ %@", [mode modeName], exe, [mode modeFlag], [mode modeExecutionDescription]];
        printf("%s\n", modeDescription.UTF8String);
    }

    printf("%s", "\n");

    for (Class mode in modes) {
        NSUInteger count = [mode modeExtraArgumentsSectionCount];
        if (count > 0) {
            for (NSUInteger i = 0; i < count; i++) {
                NSString *name = [mode modeExtraArgumentsSectionName:i];
                NSArray<NSString *> *items = [mode modeExtraArgumentsSectionDescriptions:i];
                printf("%s:\n-------------------------------\n", name.UTF8String);
                for (NSString *item in items) {
                    printf("\t%s\n", item.UTF8String);
                }
                printf("%s", "\n");
            }
        }
    }
}

NSArray<Class> *NOZCLI_allModes()
{
    return @[
             [NOZCLIMethodMode class],
             [NOZCLIDumpMode class],
             [NOZCLICompressMode class],
             [NOZCLIDecompressMode class],
             [NOZCLIZipMode class],
             [NOZCLIUnzipMode class],
             ];
}

void NOZCLI_printError(NSError *error)
{
    NSString *errorString = [NSString stringWithFormat:@"ERROR: %@", error];
    printf("%s\n", errorString.UTF8String);
}

NSString *NOZCLI_normalizedPath(NSString *envPath, NSString *fileArg)
{
    if (fileArg) {
        if ([fileArg hasPrefix:@"/"]) {
            // leave as-is
        } else if ([fileArg hasPrefix:@"~"]) {
            fileArg = [fileArg stringByExpandingTildeInPath];
        } else {
            fileArg = [envPath stringByAppendingPathComponent:fileArg];
        }
    }

    return fileArg;
}

double NOZCLI_computeCompressionRatio(long long uncompressedBytes, long long compressedBytes)
{
    double ratio;
    if (uncompressedBytes > 0 && compressedBytes > 0) {
        ratio = ((double)uncompressedBytes / (double)compressedBytes);
    } else {
        ratio = NAN;
    }
    return ratio;
}

void NOZCLI_registerExtraEncoders(void)
{
    NOZCompressionLibrary *library = [NOZCompressionLibrary sharedInstance];

    if ([NOZXAppleCompressionCoder isSupported]) {
        // Apple's Compression Lib is only supported on iOS 9+ and Mac OS X 10.11+

        // LZMA
        [library setEncoder:[NOZXAppleCompressionCoder encoderWithAlgorithm:COMPRESSION_LZMA]
                  forMethod:NOZCompressionMethodLZMA];
        [library setDecoder:[NOZXAppleCompressionCoder decoderWithAlgorithm:COMPRESSION_LZMA]
                  forMethod:NOZCompressionMethodLZMA];

        // LZ4
        [library setEncoder:[NOZXAppleCompressionCoder encoderWithAlgorithm:COMPRESSION_LZ4]
                  forMethod:(NOZCompressionMethod)COMPRESSION_LZ4];
        [library setDecoder:[NOZXAppleCompressionCoder decoderWithAlgorithm:COMPRESSION_LZ4]
                  forMethod:(NOZCompressionMethod)COMPRESSION_LZ4];

        // Apple LZFSE - the new hotness for compression from Apple
        [library setEncoder:[NOZXAppleCompressionCoder encoderWithAlgorithm:COMPRESSION_LZFSE]
                  forMethod:(NOZCompressionMethod)COMPRESSION_LZFSE];
        [library setDecoder:[NOZXAppleCompressionCoder decoderWithAlgorithm:COMPRESSION_LZFSE]
                  forMethod:(NOZCompressionMethod)COMPRESSION_LZFSE];
    }

    [library setEncoder:[NOZXBrotliCompressionCoder encoder]
              forMethod:kMethodBrotli];
    [library setDecoder:[NOZXBrotliCompressionCoder decoder]
              forMethod:kMethodBrotli];

    [library setEncoder:[NOZXZStandardCompressionCoder encoder]
              forMethod:NOZCompressionMethodZStandard];
    [library setDecoder:[NOZXZStandardCompressionCoder decoder]
              forMethod:NOZCompressionMethodZStandard];
}

NSArray<MethodInfo *> *NOZCLI_allMethods()
{
    NSMutableArray<MethodInfo *> *methods = [[NSMutableArray alloc] init];
    [methods addObjectsFromArray:NOZCLI_allDefaultMethods()];
    [methods addObjectsFromArray:NOZCLI_allExtendedMethods()];
    [methods addObjectsFromArray:NOZCLI_allUnsupportedMethods()];
    return methods;
}

MethodInfo *NOZCLI_lookupMethod(NOZCompressionMethod method)
{
    for (MethodInfo *info in NOZCLI_allMethods()) {
        if (info.method == method) {
            return info;
        }
    }
    return nil;
}

MethodInfo *NOZCLI_lookupMethodByName(NSString *method)
{
    for (MethodInfo *info in NOZCLI_allMethods()) {
        if ([info.name isEqualToString:method]) {
            return info;
        }
    }
    return nil;
}

NSArray<MethodInfo *> *NOZCLI_allUnsupportedMethods()
{
    NSMutableArray<MethodInfo *> *methods = [[NSMutableArray alloc] init];

    UNS_ADD_METHOD(methods, @"Shrink", NOZCompressionMethodShrink);
    UNS_ADD_METHOD(methods, @"Reduce #1", NOZCompressionMethodReduce1);
    UNS_ADD_METHOD(methods, @"Reduce #2", NOZCompressionMethodReduce2);
    UNS_ADD_METHOD(methods, @"Reduce #3", NOZCompressionMethodReduce3);
    UNS_ADD_METHOD(methods, @"Reduce #4", NOZCompressionMethodReduce4);
    UNS_ADD_METHOD(methods, @"Implode", NOZCompressionMethodImplode);
    UNS_ADD_METHOD(methods, @"*PKWare Reserved*", NOZCompressionMethodReserved7);
    UNS_ADD_METHOD(methods, @"Deflate", NOZCompressionMethodDeflate);
    UNS_ADD_METHOD(methods, @"Deflate64", NOZCompressionMethodDeflate64);
    UNS_ADD_METHOD(methods, @"IBM TERSE (old)", NOZCompressionMethodIBMTERSEOld);
    UNS_ADD_METHOD(methods, @"*PKWare Reserved*", NOZCompressionMethodReserved11);
    UNS_ADD_METHOD(methods, @"BZip2", NOZCompressionMethodBZip2);
    UNS_ADD_METHOD(methods, @"*PKWare Reserved*", NOZCompressionMethodReserved13);
    UNS_ADD_METHOD(methods, @"LZMA", NOZCompressionMethodLZMA);
    UNS_ADD_METHOD(methods, @"*PKWare Reserved*", NOZCompressionMethodReserved15);
    UNS_ADD_METHOD(methods, @"*PKWare Reserved*", NOZCompressionMethodReserved16);
    UNS_ADD_METHOD(methods, @"*PKWare Reserved*", NOZCompressionMethodReserved17);
    UNS_ADD_METHOD(methods, @"IBM TERSE (new)", NOZCompressionMethodIBMTERSENew);
    UNS_ADD_METHOD(methods, @"LZ77", NOZCompressionMethodLZ77);
    UNS_ADD_METHOD(methods, @"*Deprecated*", NOZCompressionMethodDeprecated20);

    UNS_ADD_METHOD(methods, @"MP3", NOZCompressionMethodMP3);
    UNS_ADD_METHOD(methods, @"XZ", NOZCompressionMethodXZ);
    UNS_ADD_METHOD(methods, @"JPEG", NOZCompressionMethodJPEG);
    UNS_ADD_METHOD(methods, @"WAV Pack", NOZCompressionMethodWAVPack);
    UNS_ADD_METHOD(methods, @"PPM v1 rev1", NOZCompressionMethodPPMv1rev1);
    UNS_ADD_METHOD(methods, @"AE-x Encrypted", NOZCompressionMethodAEXEncryption);

    return methods;
}

NSArray<MethodInfo *> *NOZCLI_allDefaultMethods()
{
    NSMutableArray<MethodInfo *> *methods = [[NSMutableArray alloc] init];

    [methods addObject:METHOD(@"None", NOZCompressionMethodNone, 0, 0, nil, nil, YES)];
    ADD_METHOD(methods, @"Deflate", NOZCompressionMethodDeflate, YES);

    return methods;
}

NSArray<MethodInfo *> *NOZCLI_allExtendedMethods()
{
    NSMutableArray<MethodInfo *> *methods = [[NSMutableArray alloc] init];

    ADD_METHOD(methods, @"ZStandard", NOZCompressionMethodZStandard, NO);
    ADD_METHOD(methods, @"Brotli", kMethodBrotli, NO); /* TODO: Apple is adding Brotli to iOS 15 (COMPRESSION_BROTLI) */

    if ([NOZXAppleCompressionCoder isSupported]) {
        ADD_METHOD(methods, @"LZMA", NOZCompressionMethodLZMA, NO);
        ADD_METHOD(methods, @"LZ4", (NOZCompressionMethod)COMPRESSION_LZ4, NO);
        ADD_METHOD(methods, @"LZFSE", (NOZCompressionMethod)COMPRESSION_LZFSE, NO);
    }

    return methods;
}

BOOL NOZCLI_registerMethodToNumberMap(NSDictionary<NSString *, NSNumber *> * __nullable methodToNumberMap)
{
    NOZCompressionLibrary *lib = [NOZCompressionLibrary sharedInstance];
    NSMutableDictionary<NSNumber *, MethodInfo *> *methodMap = [[NSMutableDictionary alloc] init];
    for (NSString *methodName in methodToNumberMap.allKeys) {
        MethodInfo *methodInfo = NOZCLI_lookupMethodByName(methodName);
        if (!methodInfo) {
            printf("No such method name: %s\n", methodName.UTF8String);
            return NO;
        }
        NSNumber *methodNumber = methodToNumberMap[methodName];
        methodMap[methodNumber] = methodInfo;
    }
    for (NSNumber *methodNumber in methodMap.allKeys) {
        MethodInfo *methodInfo = methodMap[methodNumber];
        [lib setEncoder:methodInfo.encoder
              forMethod:methodNumber.unsignedShortValue];
        [lib setDecoder:methodInfo.decoder
              forMethod:methodNumber.unsignedShortValue];
    }
    return YES;
}
